### PR TITLE
Text DSL (again)

### LINKED ISSUE #3: Text DSL
## Intro:
Text in Minecraft can be a dense and hard to read or debug system when significant amounts of styling are involved, often resulting in a similar problem to [brigadier](https://github.com/QuiltMC/quilt-kotlin-libraries/issues/1) in which multiple, nested builders create difficult to style code. This also makes the creation of text a notable target for a Kotlin DSL system.

## Existing work:
- [spaceserve-ekho](https://github.com/SpaceServe/spaceserve-ekho)

### PR DESCRIPTION
**Pull request by Cypher121**
Adds a StyleBuilder to create text Styles, and a buildStyle function wrapping it.

Adds a TextBuilder to create styled Text and a DSL wrapping it. TextBuilder uses StyleBuilder to keep track of style scoping functions and append text with the currently appropriate style objects.

Still needs samples.

Supersedes #16, closes #3

### CODE REVIEW

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/Color.kt [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    @Suppress("MagicNumber")
+    public constructor(red: Int, green: Int, blue: Int) : this(
+        (red.coerceIn(0, 255) shl 16) +
+                (green.coerceIn(0, 255) shl 8) +
+                blue.coerceIn(0, 255)
+    )
+
+    /** A color of red influenced by [value]. */
+    public
+    val red: Int get() = value shr 16 and 0xFF
```
Comment (Peanuuutz) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/Color.kt:84]: Not sure if I'm a fan of this line break in modifiers.

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+@DslMarker
+public annotation class StyleDsl
+
+/**
+ * A mutable style object that transforms to Minecraft [styles][Style].
+ *
+ * @author NoComment1105
+ */
+@StyleDsl
+public class StyleBuilder {
```
Comment (Peanuuutz) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt:40]: Add a constructor parameter of Style as default, defaulted to `Style.EMPTY` (same for the `buildStyle` at the bottom). Or add a `fun copyFrom(source: Style)`.

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+     * described by [getProp] and [setProp] to [newValue] for the duration of the action.
+     *
+     * The value is then reset to its original state.
+     *
+     * @author Cypher121
+     */
+    public inline fun <T> withProp(
+        newValue: T,
+        getProp: (StyleBuilder) -> T,
+        setProp: (StyleBuilder, T) -> Unit,
```
Comment (Peanuuutz) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt:68]: May be `StyleBuilder.(T) -> Unit`? (same for `getProp`)

**Review by Peanuuutz — APPROVED** [review]
lgtm

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+ * @param value The translation key of the text.
+ * @param args Any arguments to apply to the translatable text. Can be left
+ * empty for no arguments.
+ * @see StyleBuilder for action
+ *
+ * @author NoComment1105
+ */
+@TextDsl
+public fun TextBuilder.translatable(
+    value: String,
```
Comment (NoComment1105) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt:116]: Surely this and other functions don't exceed the line limit? Could they be put on the same line?

**Review by NoComment1105 — CHANGES_REQUESTED** [review]
1 thing, otherwise lgtm

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt (hunk 4 above)
Comment (Peanuuutz) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt:116]: Line limit?

Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt]: I'm leaving the NBT ones multiline cause they have a lot of arguments. Everything else moved to single-line

Comment (NoComment1105) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt]: All good

**Review by NoComment1105 — APPROVED** [review]
Lgtm

**Review by PotatoPresident — APPROVED** [review]

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    /** Whether to format the text with an underline or not. */
+    public var underlined: Boolean? = null
+        set(value) {
+            if (field != value) {
+                cachedStyle = null
+            }
+            field = value
+        }
+
+    /** A [HoverEvent] to apply to the text. */
```
Comment (SilverAndro) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt:96]: Why are only some of these doc comments inline? Can we standardize them?

**Review by SilverAndro — CHANGES_REQUESTED** [review]

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt (hunk 5 above)
Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/StyleDsl.kt:96]: Property KDoc is inline, except insertion by the look of it. Should I change insertion or make all of them multi-line?

**Review by SilverAndro — APPROVED** [review]

**Review by PotatoPresident — APPROVED** [review]

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/text/TextDsl.kt:386]: There's gotta be a better way to do this

**Review by Oliver-makes-code — COMMENTED** [review]
One comment, otherwise LGTM

**Review by Oliver-makes-code — APPROVED** [review]

**Comment by Cypher121** [comment]
An issue has been found by @gaming32 with the styled method: due to having a pretty large body, lots of inline-mangled names and being inline itself, it produces several kilobytes of bytecode whenever used. Some further testing is needed to see if that has much impact when compressed in a JAR, but if it does there are a few options:

* Leave it as is, consider the filesize cost a reasonable one (between the number of callsites per mod and compression, it probably is). Probably needs a warning in docs.

* Un-inline the method. Removes the issue, but can cause the method to create an object every time it's invoked. Worst case, gets repeatedly called in a render loop, but GC should be able to handle that.

* Add both versions and describe the difference in docs.

**Comment by SilverAndro** [comment]
> Un-inline the method. Removes the issue, but can cause the method to create an object every time it's invoked.

Does it? Im 80% sure that kotlinc can produce a lambda instance instead of a constructor if its not capturing anything which would mean no GC pressure, worth checking

**Comment by Peanuuutz** [comment]
I'd prefer not inline one.

**Comment by Cypher121** [comment]
Emphasis on "can", it won't always. That said, I'd expect code like this to be a common occurrence:

```kt
override fun render() {
    //other things here
    
    textRenderer.draw(
        buildText {
            styled(bold = true, hoverEvent = someEvent) {
                translatable("some.mod.power_stored", handler.powerPercentage)
            }
        }
    )
}
```

**Comment by Gaming32** [comment]
Additionally, in Kotlin 1.8 it will be changed so that all lambdas become indy-based.

**Comment by Cypher121** [comment]
Un-inline it is

**Comment by Gaming32** [comment]
I agree. In your example, one reference variable is captured, so the resultant object should be about 16 bytes.

**Comment by Gaming32** [comment]
I think the `styled` block should be inlined, since it now only calls the other `styled` function. If both are un-inlined, there are two lambda calls.

**Comment by Cypher121** [comment]
If you mean the `styled(Style, TextBuilder.() -> Unit)` extension, it doesn't re-wrap the action, just passes it through, so it doesn't allocate a new object. In fact, if it was inlined, the argument would have to be marked as noinline anyway cause it's passed to the other function

**Comment by Gaming32** [comment]
That's fair.

**Pull request merged**